compiler/rustc_expand/src/expand.rs

```diff
@@ -860,6 +960,144 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
+            tokens: None,
+        };
+        let input = pprust::expr_to_string(&expr);
+
+        // Get `mac` back out of `expr`.
+        let ast::Expr { kind: ExprKind::MacCall(mac), .. } = expr else { unreachable!() };
+
+        self.update_macro_stats(
+            MacroStatKind::FnLike,
+            fragment_kind,
+            span,
+            &mac.path,
+            &input,
+            fragment,
+        );
+    }
+
+    fn update_attr_macro_stats(
+        &mut self,
+        fragment_kind: AstFragmentKind,
+        span: Span,
+        path: &ast::Path,
+        attr: &ast::Attribute,
+        item: Annotatable,
+        fragment: &AstFragment,
+    ) {
+        // Does this path match `#[derive(...)]` in any of its forms? If so,
+        // ignore it because the individual derives will go through the
+        // `Invocation::Derive` handling separately.
+        let is_derive_path = *path == sym::derive
+            // ::core::prelude::v1::derive
+            || *path == [kw::PathRoot, sym::core, sym::prelude, sym::v1, sym::derive].as_slice();
+        if is_derive_path {
+            return;
+        }
+
+        // The attribute plus the item itself constitute the input, which we
+        // measure.
+        let input = format!(
+            "{}\n{}",
+            pprust::attribute_to_string(attr),
+            fragment_kind.expect_from_annotatables(iter::once(item)).to_string(),
+        );
+        self.update_macro_stats(MacroStatKind::Attr, fragment_kind, span, path, &input, fragment);
+    }
+
+    fn update_derive_macro_stats(
+        &mut self,
+        fragment_kind: AstFragmentKind,
+        span: Span,
+        path: &ast::Path,
+        fragment: &AstFragment,
+    ) {
+        // Use something like `#[derive(Clone)]` for the measured input, even
+        // though it may have actually appeared in a multi-derive attribute
+        // like `#[derive(Clone, Copy, Debug)]`.
+        let input = format!("#[derive({})]", pprust::path_to_string(path));
+        self.update_macro_stats(MacroStatKind::Derive, fragment_kind, span, path, &input, fragment);
+    }
+
+    fn update_macro_stats(
+        &mut self,
+        macro_stat_kind: MacroStatKind,
+        fragment_kind: AstFragmentKind,
+        span: Span,
+        path: &ast::Path,
+        input: &str,
+        fragment: &AstFragment,
+    ) {
+        fn lines_and_bytes(s: &str) -> (usize, usize) {
+            (s.trim_end().split('\n').count(), s.len())
+        }
+
+        // Measure the size of the output by pretty-printing it and counting
+        // the lines and bytes.
+        let path = pprust::path_to_string(path);
+        let output = fragment.to_string();
+        let (in_l, in_b) = lines_and_bytes(input);
+        let (out_l, out_b) = lines_and_bytes(&output);
+
+        // This code is useful for debugging `-Zmacro-stats`. For every
+        // invocation it prints the full input and output.
+        if false {
```

Comment:
Todo: not forget to remove this.

Reply:
I'd like to keep it. I used it a lot while writing and testing this PR, and there's a good chance it'll be needed in the future. I don't want to have to recreate it.